File: src/main/java/gruvexp/bbminigames/twtClassic/ability/Ability.java
package gruvexp.bbminigames.twtClassic.ability;

import gruvexp.bbminigames.Main;
import gruvexp.bbminigames.twtClassic.BotBows;
import gruvexp.bbminigames.twtClassic.BotBowsPlayer;
import org.bukkit.inventory.ItemStack;
import org.bukkit.scheduler.BukkitRunnable;

public class Ability {

    protected final BotBowsPlayer bp;
    protected AbilityType type;

    protected int baseCooldown; // seconds
    protected float cooldownMultiplier = 1.0f;
    private int effectiveCooldown;

    private final int hotBarSlot;
    private CooldownTimer cooldownTimer;
    private int cooldownTickRate = 20;

    public Ability(BotBowsPlayer bp, int hotBarSlot, AbilityType type) {
        this.bp = bp;
        this.hotBarSlot = hotBarSlot;
        this.type = type;
        if (type.category != AbilityCategory.DAMAGING) {
            this.baseCooldown = type.getBaseCooldown();
        }
    }

    public AbilityType getType() {
        return type;
    }

    public int getHotBarSlot() {
        return hotBarSlot;
    }

    public void setCooldownMultiplier(float multiplier) {
        this.cooldownMultiplier = multiplier;
        effectiveCooldown = (int) (baseCooldown * cooldownMultiplier);
    }

    public int getEffectiveCooldown() {
        return effectiveCooldown;
    }

    public void resetCooldown() {
        if (cooldownTimer != null) {
            cooldownTimer.cancel();
            cooldownTimer = null;
        }
    }

    public void obtain() {
        resetCooldown();
        bp.avatar.setItem(hotBarSlot, type.getAbilityItem());
    }

    public void lose() {
        if (type.category == AbilityCategory.DAMAGING) {
            bp.avatar.setItem(hotBarSlot, type.getCooldownItems()[0].clone());
        } else {
            cooldownTimer = new CooldownTimer(bp, effectiveCooldown);
            cooldownTimer.runTaskTimer(Main.getPlugin(), 0L, cooldownTickRate);
        }
    }

    public void hit() {
        if (cooldownTimer != null) {
            cooldownTimer.hit();
        }
    }

    public void use() {
        if (bp.lobby.botBowsGame != null && !bp.lobby.botBowsGame.canMove) return; // null check used when testing ability outside of match

        if (type.category == AbilityCategory.DAMAGING) {
            bp.loseWeaponAbilities();
        } else {
            cooldownTimer = new CooldownTimer(bp, effectiveCooldown);
            cooldownTimer.runTaskTimer(Main.getPlugin(), 0L, cooldownTickRate);
        }
    }

    public void setTickRate(int tickRate) {
        cooldownTickRate = tickRate;
        if (cooldownTimer != null) {
            cooldownTimer.cancel();
            int cooldownLeft = cooldownTimer.currentCooldown;

            cooldownTimer = new CooldownTimer(bp, cooldownLeft);
            cooldownTimer.runTaskTimer(Main.getPlugin(), 0L, cooldownTickRate);
        }
    }

    private ItemStack getCooldownItem(int cooldown) {
        if (cooldown > 10) {
            return type.getCooldownItems()[0].clone();
        } else if (cooldown > 5) {
            return type.getCooldownItems()[1].clone();
        } else if (cooldown > 2) {
            return type.getCooldownItems()[2].clone();
        } else {
            return type.getCooldownItems()[3].clone();
        }
    }

    public void unequip() {

    }

    public void reset() {

    }

    private class CooldownTimer extends BukkitRunnable {
        int currentCooldown;
        ItemStack cooldownItem = getCooldownItem(currentCooldown);
        private final BotBowsPlayer bp;

        private CooldownTimer(BotBowsPlayer bp, int effectiveCooldown) {
            this.bp = bp;
            currentCooldown = effectiveCooldown;
        }

        @Override
        public void run() {
            if (currentCooldown <= 0) {
                obtain();
                return;
            }

            switch (currentCooldown) {
                case 10 -> cooldownItem = type.getCooldownItems()[1].clone();
                case 5 -> cooldownItem = type.getCooldownItems()[2].clone();
                case 2 -> cooldownItem = type.getCooldownItems()[3].clone();
            }
            cooldownItem.setAmount(currentCooldown);
            bp.avatar.setItem(hotBarSlot, cooldownItem);
            currentCooldown--;
        }

        public void hit() { // when someone hits you with a bow, the cooldown wont go down until the damage cooldown is complete (when barrier blocks get removed)
            currentCooldown += BotBows.HIT_DISABLED_ITEM_TICKS / 20;
        }
    }
}

File: src/main/java/gruvexp/bbminigames/twtClassic/ability/Ability.kt
package gruvexp.bbminigames.twtClassic.ability

import gruvexp.bbminigames.Main
import gruvexp.bbminigames.twtClassic.BotBows
import gruvexp.bbminigames.twtClassic.BotBowsPlayer
import gruvexp.bbminigames.twtClassic.ability.abilities.*
import org.bukkit.inventory.ItemStack
import org.bukkit.scheduler.BukkitRunnable

open class Ability(@JvmField protected val bp: BotBowsPlayer, val hotBarSlot: Int, @JvmField val type: AbilityType) {
    @JvmField
    protected var baseCooldown: Int = 0 // seconds
    var cooldownMultiplier: Float = 1.0f
        set(value) {
            field = value
            effectiveCooldown = (baseCooldown * value).toInt()
        }
    var effectiveCooldown: Int = 0
        private set

    private var cooldownTimer: CooldownTimer? = null
    var cooldownTickRate = 20
        set(value) {
            field = value
            cooldownTimer?.let {
                it.cancel()
                cooldownTimer = CooldownTimer(bp, it.currentCooldown).apply {
                    runTaskTimer(Main.getPlugin(), 0L, value.toLong())
                }
            }
        }

    init {
        if (type.category != AbilityCategory.DAMAGING) {
            this.baseCooldown = type.baseCooldown
        }
    }

    fun resetCooldown() {
        cooldownTimer?.cancel()
        cooldownTimer = null
    }

    fun obtain() {
        resetCooldown()
        bp.avatar.setItem(hotBarSlot, type.abilityItem)
    }

    fun lose() {
        if (type.category == AbilityCategory.DAMAGING) {
            bp.avatar.setItem(hotBarSlot, type.cooldownItems[0].clone())
        } else {
            cooldownTimer = CooldownTimer(bp, effectiveCooldown).apply {
                runTaskTimer(Main.getPlugin(), 0L, cooldownTickRate.toLong())
            }
        }
    }

    fun hit() {
        cooldownTimer?.hit()
    }

    open fun use() {
        val game = bp.lobby.botBowsGame ?: return
        if (game.canMove) return  // null check used when testing ability outside of match

        if (type.category == AbilityCategory.DAMAGING) {
            bp.loseWeaponAbilities()
        } else {
            cooldownTimer = CooldownTimer(bp, effectiveCooldown).apply {
                runTaskTimer(Main.getPlugin(), 0L, cooldownTickRate.toLong())
            }
        }
    }

    fun registerSuccess() = bp.lobby.botBowsGame!!.matchResult.registerAbilitySuccess(bp, type)
    fun registerFail() = bp.lobby.botBowsGame!!.matchResult.registerAbilityFail(bp, type)

    private fun getCooldownItem(cooldown: Int) = when {
        cooldown > 10 -> type.cooldownItems[0].clone()
        cooldown > 5 ->  type.cooldownItems[1].clone()
        cooldown > 2 ->  type.cooldownItems[2].clone()
        else ->          type.cooldownItems[3].clone()
    }

    open fun unequip() {
    }

    open fun reset() { // removing things gracefully (eg igniting creeper, remove effects etc)
    }

    open fun destroy() { // removing everything by force (removing all entities without any effect)
        resetCooldown()
    }

    private inner class CooldownTimer(private val bp: BotBowsPlayer, var currentCooldown: Int) : BukkitRunnable() {
        var cooldownItem: ItemStack = getCooldownItem(currentCooldown)

        override fun run() {
            if (currentCooldown <= 0) {
                obtain()
                return
            }

            when (currentCooldown) {
                10 -> cooldownItem = type.cooldownItems[1].clone()
                5 ->  cooldownItem = type.cooldownItems[2].clone()
                2 ->  cooldownItem = type.cooldownItems[3].clone()
            }
            cooldownItem.amount = currentCooldown
            bp.avatar.setItem(hotBarSlot, cooldownItem)
            currentCooldown--
        }

        fun hit() { // when someone hits you with a bow, the cooldown wont go down until your invulnerability period is over
            currentCooldown += BotBows.HIT_DISABLED_ITEM_TICKS / 20
        }
    }

    companion object {
        fun create(type: AbilityType, bp: BotBowsPlayer, slot: Int): Ability = when (type) {
            AbilityType.ENDER_PEARL -> object : Ability(bp, slot, AbilityType.ENDER_PEARL) {
                override fun use() = registerSuccess()
            }
            AbilityType.RADAR -> Radar(bp, slot)
            AbilityType.SPLASH_BOW -> SplashBow(bp, slot)
            AbilityType.THUNDER_BOW -> ThunderBow(bp, slot)
            AbilityType.LONG_ARMS -> LongArms(bp, slot)
            AbilityType.SALMON_SLAP -> SalmonSlap(bp, slot)
            AbilityType.BUBBLE_JET -> BubbleJet(bp, slot)
            AbilityType.CREEPER_TRAP -> CreeperTrap(bp, slot)
            AbilityType.BABY_POTION -> BabyPotion(bp, slot)
            AbilityType.LINGERING_POTION -> LingeringPotionTrap(bp, slot)
            AbilityType.CHARGE_POTION -> ChargePotion(bp, slot)
            AbilityType.KARMA_POTION -> KarmaPotion(bp, slot)
            AbilityType.LASER_TRAP -> LaserTrap(bp, slot)
        }
    }
}
